README
Games-Cards-Pair

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

SUPPORT AND DOCUMENTATION

After  installing,   you can  find documentation for this module with the perldoc
command.

    perldoc Games::Cards::Pair

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Cards-Pair

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Games-Cards-Pair

    CPAN Ratings
        http://cpanratings.perl.org/d/Games-Cards-Pair

    Search CPAN
        http://search.cpan.org/dist/Games-Cards-Pair/

LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2016 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changes
Revision history for Games::Cards::Pair

0.15  Thu Oct 13 12:10:00 2016
      - Improved type definitions for 'ZeroOrOne', 'Suit' and 'Value'.

0.14  Sun Aug 14 08:10:00 2016
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.13  Thu Mar 24 15:25:00 2016
      - Added key 'provides' to the Makefile.PL script.
      - Updated copyright year information.

0.12  Thu Oct 01 10:10:00 2015
      - Added $AUTHORITY to each package of the distribution.
      - Updated the section "DESCRIPTION" in the pod document.

0.11  Sat May 30 21:15:00 2015
      - Made the game interactive (long overdue).
      - Renamed game script play-pelmanism.
      - Added section "SYNOPSIS" to the pod document.
      - Updated section "DESCRIPTION" to respect the above changes.
      - Added method init(), is_valid_card_count() and _pick().
      - Updated Games::Cards::Pair::Card::as_string().
      - Updated the card value as 1-10,A,J,K,Q.
      - Updated the card suite as C,D,H,S.

0.10  Mon Mar 09 13:55:30 2015
      - Added LICENSE file.

0.09  Sat Jan 17 00:55:30 2015
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.08  Mon Oct 20 07:15:30 2014
      - Removed the reference of 'Mouse' from the package Games::Cards::Pair::Card.

0.07  Sun Oct 19 15:40:15 2014
      - Corrected the typo in the pod document of Games::Cards::Pair::Params package.

0.06  Sun Oct 19 15:19:20 2014
      - Added the missing default 0 to 'count'.

0.05  Sun Oct 19 13:23:52 2014
      - Restructured the code using 'Moo'.
      - Added section 'Repository' to the pod document.

0.04  Sun Oct 19 09:37:20 2014
      - Removed the use of 'Carp' and 'Readonly'.
      - Added Copyright year details to the pod document.
      - Slight style changes overall.

0.03  Sun Oct 19 08:50:15 2014
      - Removed Build.PL file.
      - Converted file to unix format.

0.02  Fri Jul 27 11:35:27 2012
      - Added dependency check for Attribute::Memoize module.

0.01  Thu Jul 26 12:00:00 2012
      - First version, released.


lib/Games/Cards/Pair.pm
package Games::Cards::Pair;

$Games::Cards::Pair::VERSION   = '0.15';
$Games::Cards::Pair::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Games::Cards::Pair - Interface to the Pelmanism (Pair) Card Game.

=head1 VERSION

Version 0.15

=cut

use 5.006;
use Data::Dumper;

use Attribute::Memoize;
use List::Util qw(shuffle);
use List::MoreUtils qw(first_index);
use Term::Screen::Lite;
use Types::Standard qw(Int);
use Games::Cards::Pair::Params qw(ZeroOrOne);
use Games::Cards::Pair::Card;

use Moo;
use namespace::clean;

use overload ('""' => \&as_string);

has 'cards'     => (is => 'rw', default => sub { 12 });
has 'bank'      => (is => 'rw');
has 'seen'      => (is => 'rw');
has 'board'     => (is => 'rw');
has 'available' => (is => 'ro');
has 'screen'    => (is => 'ro', default => sub { Term::Screen::Lite->new; });
has 'count'     => (is => 'rw', isa => Int,       default => sub { 0 });
has 'debug'     => (is => 'rw', isa => ZeroOrOne, default => sub { 0 });

=head1 DESCRIPTION

A single-player game of Pelmanism, played with minimum of 12 cards and maximum up
to 54 cards. Depending on number of cards choosen the user, it prepares the game.

Cards picked up from the collection comprises each of the thirteen values (2,3,4,
5,6,7,8,9,10,Queen,King,Ace and Jack) in  each of the four suits (Clubs,Diamonds,
Hearts and Spades) plus two jokers. The Joker will not have any suit.

=head1 SYNOPSIS

Below is the working code  for the Pelmanism game using the L<Games::Cards::Pair>
package. The game script C<play-pelmanism> is supplied with the distribution  and
on install is available to play with.

    use strict; use warnings;
    use Games::Cards::Pair;

    $|=1;

    $SIG{'INT'} = sub { print {*STDOUT} "\n\nCaught Interrupt (^C), Aborting\n"; exit(1); };

    my $game = Games::Cards::Pair->new({debug => 1});

    my ($count);
    do {
        print {*STDOUT} "How many cards do you want to play with? (min:12, max:54): ";
        $count = <STDIN>;
        chomp($count);
    } until ($game->is_valid_card_count($count));

    $game->cards($count);

    my ($response);
    do {
        $game->init;
        print {*STDOUT} $game->get_board;
        print {*STDOUT} "You have 30 seconds to memorise the card positions.\n";
        sleep 30;
        $game->screen->clear;

        do {
            my ($cards);
            print {*STDOUT} $game->as_string(1);
            do {
                print {*STDOUT} "Pick pair of cards: ";
                $cards = <STDIN>;
                chomp($cards);
            } until ($cards =~ /^\d+\,\d+$/);

            $game->play($cards);

        } until ($game->is_over);

        print "\n\nTotal moves [" . $game->count . "]\n";

        do {
            print {*STDOUT} "Do you wish to continue (Y/N)? ";
            $response = <STDIN>;
            chomp($response);
        } until (defined $response && ($response =~ /^[Y|N]$/i));

    } until ($response =~ /^N$/i);

    print {*STDOUT} "Thank you.\n";

Once it is installed, it can be played on a terminal/command window  as below:

    $ play-pelmanism

=head1 METHODS

=head2 play($index)

Accepts comma separated card indices and play the game.

=cut

sub play {
    my ($self, $index) = @_;

    my ($card, $new) = $self->_pick($index);

    if ($new->equal($card)) {
        $self->_process($new, $card);
    }
    else {
        $self->{deck}->{$new->index}  = $new;
        $self->{deck}->{$card->index} = $card;
    }
}

=head2 is_over()

Returns 1 or 0 depending if the deck is empty or not.

=cut

sub is_over {
    my ($self) = @_;

    return (scalar(@{$self->{available}}) == 0);
}

=head2 get_board()

Return game board with hidden card, showing only the card index.

=cut

sub get_board {
    return $_[0]->as_string(0);
}

=head2 is_valid_card_count($count)

Valid card count is any number between 12 and 54 (both inclusive). Also it should
be a multiple of 4.

=cut

sub is_valid_card_count {
    my ($self, $count) = @_;

    return (defined $count
            && ($count =~ /^\d+$/)
            && ($count >=12 || $count <= 54)
            && ($count % 4 == 0));
}

=head2 init()

Shuffles the pack and then pick required number of cards.

=cut

sub init {
    my ($self) = @_;

    my $cards = [];
    $self->{available} = [];
    my $i = $self->cards / 4;
    foreach my $suit (qw(C D H S)) {
        my $j = 1;
        foreach my $value (qw(A 2 3 4 5 6 7 8 9 10 J Q K)) {
            next if ($j > $i);
            push @$cards, Games::Cards::Pair::Card->new({ suit => $suit, value => $value });
            $j++;
        }
    }

    # Adding two Jokers to the Suit.
    push @$cards, Games::Cards::Pair::Card->new({ value => 'Joker' });
    push @$cards, Games::Cards::Pair::Card->new({ value => 'Joker' });
    push @{$self->{available}}, $_ for (0..$self->cards-1);

    # Index card after shuffle.
    $self->_index($cards);
}

=head2 get_matched_pairs()

Returns all the matching pair, if any found, from the bank.

=cut

sub get_matched_pairs {
    my ($self) = @_;

    my $string = '';
    foreach (@{$self->{bank}}) {
        $string .= sprintf("%s %s\n", $_->[0], $_->[1]);
    }

    return $string;
}

=head2 as_string()

Returns deck arranged as 4 in a row blocks. This is overloaded as string context.

=cut

sub as_string {
    my ($self, $hide) = @_;

    my $deck = '';
    foreach my $i (1..$self->cards) {
        my $card = $self->{deck}->{$i-1};
        my $c = '     ';
        if (defined $card) {
            if ($hide) {
                $c = $i;
            }
            else {
                $c = $card->as_string;
            }
        }

        $deck .= sprintf("[ %5s ]", $c);
        $deck .= "\n" if ($i % 4 == 0);
    }

    return $deck;
}

#
#
# PRIVATE METHODS

sub _pick {
    my ($self, $index) = @_;

    $self->{count}++;

    my ($i, $j) = split /\,/, $index, 2;
    --$i; --$j;
    my $c1 = $self->{deck}->{$i};
    die "ERROR: Invalid card received [$i].\n" unless defined $c1;

    my $c2 = $self->{deck}->{$j};
    die "ERROR: Invalid card received [$j].\n" unless defined $c2;

    push @{$self->{seen}}, $c1, $c2;

    return ($c1, $c2);
}

sub _save {
    my ($self, @cards) = @_;

    die("ERROR: Expecting atleast a pair of cards.\n") unless (scalar(@cards) == 2);

    push @{$self->{bank}}, [@cards];
}

sub _process {
    my ($self, $card, $new) = @_;

    $self->{deck}->{$new->index}  = undef;
    $self->{deck}->{$card->index} = undef;
    $self->_save($card, $new);

    my $index = first_index { $_ == $new->index } @{$self->{available}};
    splice(@{$self->{available}}, $index, 1) if ($index != -1);

    $index = first_index { $_ == $card->index } @{$self->{available}};
    splice(@{$self->{available}}, $index, 1) if ($index != -1);
}

sub _index {
    my ($self, $cards) = @_;

    $cards = [shuffle @{$cards}];
    my $index  = 0;
    foreach my $card (@{$cards}) {
        $card->index($index);
        $self->{deck}->{$index} = $card;
        $index++;
        last if ($self->cards == $index);
    }
}

sub _draw {
    my ($self) = @_;

    my @random = shuffle(@{$self->{available}});
    my $index  = shift @random;

    return $self->{deck}->{$index} if defined $index;
    return;
}

sub _seen :Memoize {
    my ($self, $card) = @_;

    my $index = 0;
    foreach (@{$self->{seen}}) {
        if ($card->equal($_)) {
            splice(@{$self->{seen}}, $index, 1);
            return $_;
        }
        $index++;
    }

    return;
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Games-Cards-Pair>

=head1 BUGS

Please report any bugs / feature requests to C<bug-games-cards-pair at rt.cpan.org>
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Cards-Pair>.
I will be notified and then you'll automatically be  notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Cards::Pair

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Cards-Pair>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Cards-Pair>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Cards-Pair>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Cards-Pair/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Games::Cards::Pair


lib/Games/Cards/Pair/Card.pm
package Games::Cards::Pair::Card;

$Games::Cards::Pair::Card::VERSION   = '0.15';
$Games::Cards::Pair::Card::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Games::Cards::Pair::Card - Object representation of a card.

=head1 VERSION

Version 0.15

=cut

use 5.006;
use Data::Dumper;
use Types::Standard qw(Int);
use Games::Cards::Pair::Params qw(Value Suit);

use Moo;
use namespace::clean;

use overload ( '""'  => \&as_string );

has 'index' => (is => 'rw', isa => Int );
has 'suit'  => (is => 'ro', isa => Suit);
has 'value' => (is => 'ro', isa => Value, required => 1);

=head1 DESCRIPTION

Only for internal use of Games::Cards::Pair class. Avoid using it directly.

=cut

sub BUILDARGS {
    my ($class, $args) = @_;

    if (defined($args->{'value'}) && ($args->{'value'} =~ /Joker/i)) {
        die("Attribute (suit) is NOT required for Joker.") if defined $args->{'suit'};
    }
    else {
        die("Attribute (suit) is required.") unless defined $args->{'suit'};
    }

    return $args;
};

=head1 METHODS

=head2 equal()

Returns 1 or 0 depending whether the two cards are same in value or one of them is a Joker.

    use strict; use warnings;
    use Games::Cards::Pair::Card;

    my ($card1, $card2);
    $card1 = Games::Cards::Pair::Card->new({ suit => 'Clubs',    value => '2' });
    $card2 = Games::Cards::Pair::Card->new({ suit => 'Diamonds', value => '2' });
    print "Card are the same.\n" if $card1->equal($card2);

    $card2 = Games::Cards::Pair::Card->new({ value => 'Joker' });
    print "Card are the same.\n" if $card1->equal($card2);

=cut

sub equal {
    my ($self, $other) = @_;

    return 0 unless (defined($other) && (ref($other) eq 'Games::Cards::Pair::Card'));

    return 1
        if ((defined($self->{value}) && ($self->{value} =~ /Joker/i))
            ||
            (defined($other->{value}) && ($other->{value} =~ /Joker/i))
            ||
            (defined($self->{value}) && (defined($other->{value}) && (lc($self->{value}) eq lc($other->{value})))));

    return 0;
}

=head2 as_string()

Returns the card object in readable format. This is overloaded as string context for printing.

    use strict; use warnings;
    use Games::Cards::Pair::Card;

    my $card = Games::Cards::Pair::Card->new({ suit => 'Clubs', value => '2' });
    print "Card: $card\n";
    # or
    print "Card: " . $card->as_string() . "\n";

=cut

sub as_string {
    my ($self) = @_;

    return sprintf("%4s%s", $self->value, $self->suit) if defined $self->suit;

    return sprintf("%5s", $self->value);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Games-Cards-Pair>

=head1 BUGS

Please report any bugs / feature requests to C<bug-games-cards-pair at rt.cpan.org>,or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Cards-Pair>.I will
be notified, & then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Cards::Pair::Card

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Cards-Pair>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Cards-Pair>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Cards-Pair>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Cards-Pair/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Games::Cards::Pair::Card


lib/Games/Cards/Pair/Params.pm
package Games::Cards::Pair::Params;

$Games::Cards::Pair::Params::VERSION   = '0.15';
$Games::Cards::Pair::Params::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Games::Cards::Pair::Params - Placeholder for parameters for Games::Cards::Pair.

=head1 VERSION

Version 0.15

=cut

use 5.006;
use strict; use warnings;
use Data::Dumper;

use Type::Library -base, -declare => qw(ZeroOrOne Suit Value);
use Types::Standard qw(Str);
use Type::Utils;

my $SUITS  = { 'C' => 1, 'D' => 1, 'H' => 1, 'S' => 1 };
my $VALUES = { '2' => 1, '3' => 1, '4' => 1, '5' => 1,
               '6' => 1, '7' => 1, '8' => 1, '9' => 1,
               '10'=> 1, 'A' => 1, 'J' => 1, 'Q' => 1,
               'K' => 1, 'Joker' => 1 };

declare 'ZeroOrOne',
    as Str,
    where   { /^[0|1]$/ },
    message { "ERROR: Only 0 or 1 allowed." };

declare 'Suit',
    as Str,
    where   { exists $SUITS->{ucfirst(lc($_[0]))} },
    message { "isa check for 'suit' failed." };

declare 'Value',
    as Str,
    where   { exists $VALUES->{ucfirst(lc($_[0]))} },
    message { "isa check for 'value' failed." };

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Games-Cards-Pair>

=head1 BUGS

Please report any bugs or feature requests to C<bug-games-cards-pair at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Cards-Pair>.
I will be notified and then you'll automatically be notified of  progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Cards::Pair::Params

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Cards-Pair>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Cards-Pair>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Cards-Pair>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Cards-Pair/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Games::Cards::Pair::Params


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Games::Cards::Pair',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Games/Cards/Pair.pm',
    ABSTRACT_FROM      => 'lib/Games/Cards/Pair.pm',
    LICENSE            => 'artistic_2',
    EXE_FILES          => [ 'script/play-pelmanism' ],
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Attribute::Memoize' => 1.01,
        'List::Util'         => 1.19,
        'List::MoreUtils'    => 0.33,
        'Term::Screen::Lite' => 0,
        'Moo'                => '2.000000',
        'namespace::clean'   => '0.24',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Games-Cards-Pair-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Games::Cards::Pair'         => { file => 'lib/Games/Cards/Pair.pm'       , version => '0.15' },
            'Games::Cards::Pair::Card'   => { file => 'lib/Games/Cards/Pair/Card.pm'  , version => '0.15' },
            'Games::Cards::Pair::Params' => { file => 'lib/Games/Cards/Pair/Params.pm', version => '0.15' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Games-Cards-Pair.git',
                web  => 'https://github.com/manwar/Games-Cards-Pair',
            },
        }})
     : ()
    ),
);
